Fail when not specifying number of decomposition components to plot

mwalls opened the pull request:
to stop it exploding when you forget to put a number

mwalls committed:
set default number of components to 3 in decomposition

francisco-dlp commented:
Wouldn't it have been easier to set the number in [plot_decomposition_factors](https://github.com/hyperspy/hyperspy/blob/RELEASE_next_minor/hyperspy/signal.py#L740) and plot_decomposition_loading](https://github.com/hyperspy/hyperspy/blob/RELEASE_next_minor/hyperspy/signal.py#L863)? e.g. 

```python

    def plot_decomposition_loadings(self,
                                    comp_ids=3,
                                    calibrate=True,
                                    same_window=None,
                                    comp_label='Decomposition loading',
                                    with_factors=False,
                                    cmap=plt.cm.gray,
                                    no_nans=False,
                                    per_row=3):
```

mwalls commented:
Yes of course. As you know, my programming style is to change random lines of code till it does what I want.  But I have two questions:
1) Does your way mean that the default is changed to 3 but if someone (for what reason I don't know) actually enters "None" they will still get all of them?
2) Ok we'll use your method but I have to do it, right? (If I do the pull request I've got to make the changes?). Still a bit new to this...

mwalls commented:
Also, if anyone has strong feelings about it being something other than 3 components, I'm not going to fight about it.

francisco-dlp commented:
I am actually wondering if it is not better not to set any default and let the function fail if the number of components is not specified. to do that simply remove the ``None`` from the function keyword argument.

You are not obliged to do it, but, since it's your idea and you've already at it...

The easiest way to proceed is to revert the changes, commit the new version and push e.g.

```bash

git revert 1535cd6

### change the code ###

git commit -a -m "Set default number of components for plotting"
git push

```

mwalls committed:
Revert "set default number of components to 3 in decomposition"

This reverts commit 1535cd6dda7ec69559ab7f674ffaf5cfe344a567.

ericpre commented:
+1 on letting fail the function if the number of components is not specified. I don't think that there is a sensible default.
If there is one thing that I remember of my early days of HyperSpy is the frustration just after realising that I forgot to specify the number of components... Now, I am very well vaccinated with that one! :D

mwalls committed:
generate error if no number of components for decomposition plots

generate error if no number of components for decomposition plots
(instead of the default of 3 originally suggested).

mwalls commented:
Ive just tried to do it with the error message but can't tell if you're all seeing it. Sorry.

ericpre commented on hyperspy/signal.py:
```diff
@@ -374,7 +374,7 @@ def _plot_factors_or_pchars(self, factors, comp_ids=None,
         else:
             return f
 
-    def _plot_loadings(self, loadings, comp_ids=None, calibrate=True,
```
This was not necessary to change the definition of this function, but it does not have any consequences. It just makes the code a bit less let consistence with the definition of `_plot_factors_or_pchars`.

ericpre commented:
Yes, this is good and it could be merged: there is just a minor comment, but it is not really important.
The tests failing in travis are not related to this PR.

mwalls commented:
ok by me. Can't remember if I'm allowed to close one. Will try now.

mwalls commented:
Looks like I shouldn't have done that.

ericpre commented:
Yes, please just wait that somebody merges it!
